### DISCUSSION TITLE
Idiom for `Pytree` impl for wrapper class

### DISCUSSION THREAD

**Discussion by femtomc** [jax-ml/jax#12358] [Q&A] [answered]
I'm trying to implement `Pytree` for a hashable variant of `dict` so I can use dictionaries with `static_argnums` in `jax.jit`:

```python
class HashableDict(dict):
    def __key(self):
        return tuple((k, self[k]) for k in sorted(self))

    def __hash__(self):
        return hash(self.__key())

    def __eq__(self, other):
        return self.__key() == other.__key()
```

This doesn't behave like the builtin `dict` variant:

```python
import jax.tree_util as jtu

d = {"x": 5.0, "y": 10.0}


def inner(v):
    print(v)
    return v


jtu.tree_map(inner, d)


class HashableDict(dict):
    def __key(self):
        return tuple((k, self[k]) for k in sorted(self))

    def __hash__(self):
        return hash(self.__key())

    def __eq__(self, other):
        return self.__key() == other.__key()


d = HashableDict({"x": 5.0, "y": 10.0})

jtu.tree_map(inner, d)
```

results in:

```
5.0
10.0
{'x': 5.0, 'y': 10.0}
```

In the builtin `dict`, the leaves are the values -- in the wrapper, the entire inner dictionary is a leaf.

Is there a convenient way to passthrough the `Pytree` methods to the inner dictionary (?)

**Comment by femtomc** [comment, marked as answer]
Here's how I solved this:

```python
jtu.register_pytree_node(
    HashableDict,
    lambda x: (list(x.values()), list(x.keys())),
    lambda keys, values: HashableDict(jax.util.safe_zip(keys, values)),
)
```

By copying some code from this issue: https://github.com/google/jax/issues/4085

**Reply by jakevdp** [reply]
That seems like the right approach; it's documented here: https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees
